sdk/identity/src/token_credentials/client_certificate_credentials.rs

```diff
@@ -0,0 +1,232 @@
+    pub fn new(
+        tenant_id: String,
+        client_id: String,
+        client_certificate: String,
+        client_certificate_pass: String,
+        options: CertificateCredentialOptions,
+    ) -> ClientCertificateCredential {
+        ClientCertificateCredential {
+            tenant_id,
+            client_id,
+            client_certificate,
+            client_certificate_pass,
+            options,
+        }
+    }
+
+    fn options(&self) -> &CertificateCredentialOptions {
+        &self.options
+    }
+
+    fn sign(jwt: &str, pkey: &PKey<Private>) -> Result<Vec<u8>, ErrorStack> {
+        let mut signer = Signer::new(MessageDigest::sha256(), pkey)?;
+        signer.update(jwt.as_bytes())?;
+        signer.sign_to_vec()
+    }
+
+    fn get_thumbprint(cert: &X509) -> Result<DigestBytes, ErrorStack> {
+        let der = cert.to_der()?;
+        let digest = hash(MessageDigest::sha1(), &der)?;
+        Ok(digest)
+    }
+
+    fn as_jwt_part(part: &[u8]) -> String {
+        base64::encode_config(part, BASE64_URL_SAFE)
+    }
+}
+
+#[allow(missing_docs)]
+#[non_exhaustive]
+#[derive(Debug, thiserror::Error)]
+pub enum ClientCertificateCredentialError {
+    #[error("Failed to perform base64 decode {0}")]
+    DecodeError(base64::DecodeError),
+    #[error("Openssl operation failed {0}")]
+    OpensslError(ErrorStack),
+    #[error("HTTP operation failed {0}")]
+    ReqwestError(reqwest::Error),
+}
+
+#[derive(Deserialize, Debug, Default)]
+#[serde(default)]
+struct AadTokenResponse {
+    token_type: String,
+    expires_in: u64,
+    ext_expires_in: u64,
+    access_token: String,
+}
+
+#[async_trait::async_trait]
+impl TokenCredential for ClientCertificateCredential {
+    type Error = ClientCertificateCredentialError;
+
+    async fn get_token(&self, resource: &str) -> Result<TokenResponse, Self::Error> {
+        let options = self.options();
+        let url = &format!(
+            "{}/{}/oauth2/v2.0/token",
+            options.authority_host(),
+            self.tenant_id
+        );
+
+        let certificate = base64::decode(&self.client_certificate)
+            .map_err(ClientCertificateCredentialError::DecodeError)?;
+        let certificate = Pkcs12::from_der(&certificate)
+            .map_err(ClientCertificateCredentialError::OpensslError)?
+            .parse(&self.client_certificate_pass)
+            .map_err(ClientCertificateCredentialError::OpensslError)?;
+
+        let thumbprint = ClientCertificateCredential::get_thumbprint(&certificate.cert)
+            .map_err(ClientCertificateCredentialError::OpensslError)?;
+
+        let uuid = uuid::Uuid::new_v4();
+        let current_time = Utc::now().timestamp();
+        let expiry_time = current_time + DEFAULT_REFRESH_TIME;
+        let x5t = base64::encode(&thumbprint);
+
+        let header = match options.send_certificate_chain {
+            true => format!(
+                r#"{{"alg":"RS256","typ":"JWT", "x5t":"{}", "x5c":["{}"]}}"#,
```

Comment:
Looks like x5c can contain only one cert? A client may provide any number of intermediate certs in addition to the signing cert, in which case x5c must contain all of them.

Reply:
Uncertain if there is a change requested here?

Reply:
The "x5c" array should take one or more certificates.

Reply:
Fixed. Sending the entire chain if available.